benchmark/fs/bench-opendir.js

```diff
@@ -0,0 +1,49 @@
+'use strict';
+
+const common = require('../common');
+const fs = require('fs');
+const path = require('path');
+
+const bench = common.createBenchmark(main, {
+  n: [100],
+  dir: [ 'lib', 'test/parallel'],
+  mode: [ 'async', 'sync', 'callback' ]
+});
+
+async function main({ n, dir, mode }) {
+  const fullPath = path.resolve(__dirname, '../../', dir);
+
+  bench.start();
+
+  let counter = 0;
+  for (let i = 0; i < n; i++) {
+    if (mode === 'async') {
+      // eslint-disable-next-line no-unused-vars
+      for await (const entry of await fs.promises.opendir(fullPath))
+        counter++;
+    } else if (mode === 'callback') {
+      const dir = await fs.promises.opendir(fullPath);
+      await new Promise((resolve, reject) => {
+        function read() {
+          dir.read((err, entry) => {
+            if (err)
+              reject(err);
+            if (entry === null)
+              resolve(dir.close());
+            else
+              read();
+          });
+        }
+
+        read();
+      });
+    } {
```

Comment:
@addaleax This patch seems wrong? There should be an `else` here... I think the benchmark you ran included `readSync()` along with the callback one...

Reply:
Oops, yes … funny the linter wouldn’t complain about this? The correct benchmark results for changing to setImmediate are 

```
                                                                confidence improvement accuracy (*)    (**)   (***)
 fs/bench-opendir.js mode='callback' dir='lib' n=100                           3.40 %      ±15.95% ±21.22% ±27.62%
 fs/bench-opendir.js mode='callback' dir='test/parallel' n=100        ***    -43.16 %      ±10.70% ±14.27% ±18.66%
```

Reply:
This result seems strange to me, how is that worse than before?

Regardless though after further though I think that `nextTick` is the reasonable thing to do.

Reply:
@Fishrock123 I think it makes sense – as you noted, the `readSync()` part was run unconditionally before, so its (unchanged) performance was included in the -30 %. If we only benchmark what has been slowed down, seeing more negative impact seems about right to me.